### PR TITLE
WIP: Fix loader path

### LINKED ISSUE #1: Issues when building HDF5 with libgfortran on Mac
I'm trying to get the HDF5 conda-forge to build with the fortran bindings:

https://travis-ci.org/conda-forge/hdf5-feedstock/builds/145963897

This isn't working on MacOS X due to the following error:

```
dyld: Library not loaded: @rpath/./libgfortran.3.dylib
  Referenced from: /Users/travis/miniconda3/conda-bld/work/hdf5-1.8.17/fortran/src/./H5test_kind
  Reason: image not found
/bin/sh: line 1: 28842 Trace/BPT trap: 5       ./H5test_kind > H5fortran_detect.f90
make[2]: *** [H5fortran_detect.f90] Error 133
make[1]: *** [all-recursive] Error 1
make: *** [all-recursive] Error 1
Command failed: /bin/bash -x -e /Users/travis/miniconda3/conda-bld/work/hdf5-1.8.17/conda_build.sh
```

However, the same recipe passes on CircleCI/Linux. The main difference between the two (apart from OS) is that the CIrcleCI build used libgfortran from the defaults channel:

https://circleci.com/gh/conda-forge/hdf5-feedstock/67

I'm not sure if this is related, but I'm opening this in case any of the libgfortran feedstock maintainers have any ideas. I'll also continue to investigate this (cc @pelson @jakirkham @msarahan)

### PR DESCRIPTION
**Pull request by jakirkham**
Fixes https://github.com/conda-forge/libgfortran-feedstock/issues/1

Not sure if this actually fixes the issue or not as I haven't tested it yet. However, this is a neat trick that may improve the situation.

Basically we delete the existing `LC_RPATH`. Then we change all the existing linked library paths to point to absolute paths with the full `PREFIX`. At this point, `conda-build` picks up on the absolute paths and fixes them. It also goes back and adds the correct `LC_RPATH`. The last step would not be possible if we hadn't already removed it.

This only has an effect on local builds of this recipe and not on conda-forge builds. To ensure we don't mess up a user's `gcc` package (as all the libraries we are mucking with are hard linked to the same file in the `gcc` package), we make fresh copies of all the libraries we mess with. As a result, we can be confident that we have mucked only with our copies of these libraries for build purposes.

In short, we trick `conda-build` into fixing these libraries for us.

Caveats: As the `libgfortran` package is just part of `gcc` on OS X, there is no guarantee that our "fixed" `libgfortran` will have its files installed or `gcc` will. It depends on who "wins" (is last) in the install. Similarly this situation may present itself with `libgcc`. Sadly there is nothing we can really do about this ATM. One will need to specify `DYLD_FALLBACK_LIBRARY_PATH` for these cases. Note that the latter will be [tricky](http://apple.stackexchange.com/questions/212945/unable-to-set-dyld-fallback-library-path-in-shell-on-osx-10-11-1) for people building on 10.11. See issue ( https://github.com/conda-forge/staged-recipes/issues/913 ) for some discussion on this last point.

### CODE REVIEW

**Comment by jakirkham** [comment]
Seem to be running into an interesting issue ( https://github.com/conda/conda-build/issues/1184 ) w.r.t. the prefix.

**Comment by pelson** [comment]
I'd be happy to give this a shot. Is there a way of testing (conveniently) that at least _some_ of the changes continue to work so that we have confidence with future changes?

**Comment by jakirkham** [comment]
So I see 3 solutions that are immediately obvious to me. Though maybe there are others.
1. Build this and other things with this locally.
2. Come up with some sort of `test` branch that deploys to the `test` label.
3. More elegant form of 2 with staged deployments.

I think the right long term solution is 3 to have deployment stages as is being discussed in issue ( https://github.com/conda-forge/conda-forge.github.io/issues/204 ). Though I don't know if getting all the way there will be the right way to test this. Still having staged deployments (so that we don't catastrophically break the system when making releases) feels like the right long term solution. It allows for rapid development which is highly desirable, but minimize damage to those using that stage. Only when we are really comfortable do we release something proper to `main`.

The less elegant but more easily realized solution is 2. A form of staged deployments that would use a `test` deployment branch on this feedstock like what is proposed for `hdf5` in PR ( https://github.com/conda-forge/hdf5-feedstock/pull/32 ). Maybe we can iron out how that one will work and try applying it here.

Edit: Needless to say 1 isn't exactly convenient. So maybe we should work on one of the other two.

**Comment by jakirkham** [comment]
Do you have any thoughts on this, @mingwandroid?

**Comment by mingwandroid** [comment]
The root of all these problems is that `libgcc_ext.10.4.dylib` and `libgcc_ext.10.5.dylib` are `dylib stubs` and `install_name_tool` refuses to process them (for no good reason that I know of). https://github.com/conda/conda-build/issues/140

To fix this, since I was pushed for time, I simply added these files to `gcc-4.8.5-8/info/has_prefix` and repackaged them (rather than rebuilding from scratch .. I don't want to spend any more time on these GCC builds since I'd rather spend that time working on the new ones instead):

```
/Users/ray/mc-x64-3.5/conda-bld/gcc-4.8_1477649012852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac binary lib/libgcc_ext.10.4.dylib
/Users/ray/mc-x64-3.5/conda-bld/gcc-4.8_1477649012852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac binary lib/libgcc_ext.10.5.dylib
```

You need to rebuild any packages which have `.dylib` or `.so` files with  `LC_LOAD_DYLIB` commands where the name begins with `/Users/ray`.

**Comment by msarahan** [comment]
@jakirkham can this be closed?
